File: Sources/Models/FileNode.swift
// Disk Inventory Zed — a modern, fast, native disk usage visualizer
//
// Copyright (C) 2026 Matt Ivan
// Licensed under GPL-3.0-or-later.

import Foundation

/// An immutable snapshot of a file-system entry.
///
/// The scanner builds the complete graph off the main actor and only then publishes it.
/// Every stored property is immutable, so sharing a completed snapshot across actors is safe.
/// Keeping the node as a reference type avoids copying entire subtrees for navigation and search.
final class FileNode: Identifiable, Hashable, @unchecked Sendable {
    enum Kind: String, Codable, Sendable {
        case file
        case directory
        case package
        case symbolicLink
    }

    let id: String
    let url: URL
    let name: String
    let kind: Kind
    let isPackage: Bool
    let isSymbolicLink: Bool
    let modificationDate: Date?
    let creationDate: Date?
    let `extension`: String?
    let logicalSize: Int64
    let allocatedSize: Int64
    let children: [FileNode]
    let errorDescription: String?
    let isHardLinkDuplicate: Bool
    let totalFileCount: Int
    let totalDirectoryCount: Int

    var size: Int64 { allocatedSize }
    var isDirectory: Bool { kind == .directory }
    var isUnreadable: Bool { errorDescription != nil }
    var isLeaf: Bool { children.isEmpty }

    var formattedSize: String {
        ByteCountFormatter.string(fromByteCount: allocatedSize, countStyle: .file)
    }

    var formattedLogicalSize: String {
        ByteCountFormatter.string(fromByteCount: logicalSize, countStyle: .file)
    }

    var directoryChildren: [FileNode]? {
        let directories = children.filter(\.isDirectory)
        return directories.isEmpty ? nil : directories
    }

    var displayName: String {
        name.isEmpty ? url.path : name
    }

    var path: String {
        url.path
    }

    init(
        id: String? = nil,
        url: URL,
        name: String,
        kind: Kind,
        isPackage: Bool? = nil,
        isSymbolicLink: Bool? = nil,
        modificationDate: Date? = nil,
        creationDate: Date? = nil,
        extension: String? = nil,
        logicalSize: Int64,
        allocatedSize: Int64,
        children: [FileNode] = [],
        errorDescription: String? = nil,
        isHardLinkDuplicate: Bool = false,
        totalFileCount: Int? = nil,
        totalDirectoryCount: Int? = nil
    ) {
        self.id = id ?? url.standardizedFileURL.path
        self.url = url
        self.name = name
        self.kind = kind
        self.isPackage = isPackage ?? (kind == .package)
        self.isSymbolicLink = isSymbolicLink ?? (kind == .symbolicLink)
        self.modificationDate = modificationDate
        self.creationDate = creationDate
        self.extension = `extension`
        self.logicalSize = max(0, logicalSize)
        self.allocatedSize = max(0, allocatedSize)
        self.children = children
        self.errorDescription = errorDescription
        self.isHardLinkDuplicate = isHardLinkDuplicate
        self.totalFileCount = totalFileCount ?? (
            kind == .directory
                ? children.reduce(0) { $0 + $1.totalFileCount }
                : 1
        )
        self.totalDirectoryCount = totalDirectoryCount ?? (
            kind == .directory
                ? 1 + children.reduce(0) { $0 + $1.totalDirectoryCount }
                : 0
        )
    }

    static func == (lhs: FileNode, rhs: FileNode) -> Bool {
        lhs.id == rhs.id
    }

    func hash(into hasher: inout Hasher) {
        hasher.combine(id)
    }

    func findChild(withID targetID: String) -> FileNode? {
        var stack = [self]
        while let node = stack.popLast() {
            if node.id == targetID { return node }
            stack.append(contentsOf: node.children.reversed())
        }
        return nil
    }

    func findChild(at targetURL: URL) -> FileNode? {
        findChild(withID: targetURL.standardizedFileURL.path)
    }

    func path(to targetID: String) -> [FileNode]? {
        var stack = [self]
        var nodesByID: [String: FileNode] = [id: self]
        var parentByID: [String: String] = [:]

        while let node = stack.popLast() {
            if node.id == targetID {
                var result: [FileNode] = []
                var cursor: String? = targetID
                while let currentID = cursor,
                      let currentNode = nodesByID[currentID] {
                    result.append(currentNode)
                    cursor = parentByID[currentID]
                }
                return Array(result.reversed())
            }

            for child in node.children.reversed() {
                nodesByID[child.id] = child
                parentByID[child.id] = node.id
                stack.append(child)
            }
        }
        return nil
    }

    /// Returns a new immutable tree with the requested item removed.
    /// Only ancestors of the removed item are rebuilt.
    func removingDescendant(withID targetID: String) -> FileNode {
        guard targetID != id,
              let targetPath = path(to: targetID),
              targetPath.count > 1 else {
            return self
        }

        var replacement: FileNode?
        for ancestor in targetPath.dropLast().reversed() {
            let childToReplaceID = replacement?.id ?? targetID
            let updatedChildren = ancestor.children.compactMap { child -> FileNode? in
                guard child.id == childToReplaceID else { return child }
                return replacement
            }
            replacement = ancestor.replacingChildren(updatedChildren)
        }
        return replacement ?? self
    }

    func descendantCounts() -> (files: Int, directories: Int) {
        (totalFileCount, totalDirectoryCount)
    }

    private func replacingChildren(_ updatedChildren: [FileNode]) -> FileNode {
        let newLogicalSize = updatedChildren.reduce(Int64(0)) { $0 + $1.logicalSize }
        let newAllocatedSize = updatedChildren.reduce(Int64(0)) { $0 + $1.allocatedSize }

        return FileNode(
            id: id,
            url: url,
            name: name,
            kind: kind,
            isPackage: isPackage,
            isSymbolicLink: isSymbolicLink,
            modificationDate: modificationDate,
            creationDate: creationDate,
            extension: `extension`,
            logicalSize: isDirectory ? newLogicalSize : logicalSize,
            allocatedSize: isDirectory ? newAllocatedSize : allocatedSize,
            children: updatedChildren,
            errorDescription: errorDescription,
            isHardLinkDuplicate: isHardLinkDuplicate,
            totalFileCount: updatedChildren.reduce(0) { $0 + $1.totalFileCount },
            totalDirectoryCount: 1 + updatedChildren.reduce(0) { $0 + $1.totalDirectoryCount }
        )
    }
}

File: Windows/src/DiskInventoryZed.Core/Models/FileNode.cs
// Disk Inventory Zed - a modern, fast disk usage visualizer
// Copyright (C) 2026 Matt Ivan. Licensed under GPL-3.0-or-later.

using System.Collections.ObjectModel;

namespace DiskInventoryZed.Core.Models;

public enum FileNodeKind
{
    File,
    Directory,
    Package,
    SymbolicLink
}

/// <summary>An immutable snapshot of one file-system entry.</summary>
public sealed class FileNode : IEquatable<FileNode>
{
    private readonly ReadOnlyCollection<FileNode> _children;

    public FileNode(
        string fullPath,
        string name,
        FileNodeKind kind,
        long logicalSize,
        long allocatedSize,
        IEnumerable<FileNode>? children = null,
        DateTimeOffset? creationDate = null,
        DateTimeOffset? modificationDate = null,
        bool isSymbolicLink = false,
        bool isHardLinkDuplicate = false,
        bool hasUnverifiedHardLinks = false,
        string? issue = null,
        int? totalFileCount = null,
        int? totalDirectoryCount = null,
        bool isUnreadable = false)
    {
        FullPath = fullPath;
        Id = fullPath;
        Name = string.IsNullOrWhiteSpace(name) ? fullPath : name;
        Kind = kind;
        LogicalSize = Math.Max(0, logicalSize);
        AllocatedSize = Math.Max(0, allocatedSize);
        CreationDate = creationDate;
        ModificationDate = modificationDate;
        IsSymbolicLink = isSymbolicLink || kind == FileNodeKind.SymbolicLink;
        IsHardLinkDuplicate = isHardLinkDuplicate;
        HasUnverifiedHardLinks = hasUnverifiedHardLinks;
        IsUnreadable = isUnreadable;
        Issue = issue;

        var childArray = children?.ToArray() ?? [];
        _children = Array.AsReadOnly(childArray);
        TotalFileCount = totalFileCount ?? (IsContainer ? childArray.Sum(child => child.TotalFileCount) : 1);
        TotalDirectoryCount = totalDirectoryCount ??
            (IsContainer ? 1 + childArray.Sum(child => child.TotalDirectoryCount) : 0);
    }

    public string Id { get; }
    public string FullPath { get; }
    public string Name { get; }
    public string DisplayName => Name;
    public FileNodeKind Kind { get; }
    public bool IsDirectory => Kind == FileNodeKind.Directory;
    public bool IsPackage => Kind == FileNodeKind.Package;
    public bool IsContainer => Kind is FileNodeKind.Directory or FileNodeKind.Package;
    public bool IsSymbolicLink { get; }
    public bool IsHardLinkDuplicate { get; }
    public bool HasUnverifiedHardLinks { get; }
    public bool IsUnreadable { get; }
    public string? Issue { get; }
    public string? Extension => IsDirectory
        ? null
        : Path.GetExtension(Name).TrimStart('.') is { Length: > 0 } extension ? extension : null;
    public long LogicalSize { get; }
    public long AllocatedSize { get; }
    public string FormattedSize => Utilities.ByteSizeFormatter.Format(AllocatedSize);
    public string FormattedLogicalSize => Utilities.ByteSizeFormatter.Format(LogicalSize);
    public DateTimeOffset? CreationDate { get; }
    public DateTimeOffset? ModificationDate { get; }
    public IReadOnlyList<FileNode> Children => _children;
    public IReadOnlyList<FileNode> DirectoryChildren => _children.Where(child => child.IsDirectory).ToArray();
    public int TotalFileCount { get; }
    public int TotalDirectoryCount { get; }

    public FileNode? FindById(string targetId)
    {
        var stack = new Stack<FileNode>();
        stack.Push(this);
        while (stack.TryPop(out var node))
        {
            if (PathComparer.Equals(node.Id, targetId))
            {
                return node;
            }

            for (var index = node.Children.Count - 1; index >= 0; index--)
            {
                stack.Push(node.Children[index]);
            }
        }

        return null;
    }

    public IReadOnlyList<FileNode>? PathTo(string targetId)
    {
        var stack = new Stack<FileNode>();
        var nodes = new Dictionary<string, FileNode>(PathComparer) { [Id] = this };
        var parents = new Dictionary<string, string>(PathComparer);
        stack.Push(this);

        while (stack.TryPop(out var node))
        {
            if (PathComparer.Equals(node.Id, targetId))
            {
                var result = new List<FileNode>();
                var cursor = targetId;
                while (nodes.TryGetValue(cursor, out var current))
                {
                    result.Add(current);
                    if (!parents.TryGetValue(cursor, out var parentId))
                    {
                        break;
                    }

                    cursor = parentId;
                }

                result.Reverse();
                return result;
            }

            for (var index = node.Children.Count - 1; index >= 0; index--)
            {
                var child = node.Children[index];
                nodes[child.Id] = child;
                parents[child.Id] = node.Id;
                stack.Push(child);
            }
        }

        return null;
    }

    public bool Equals(FileNode? other) => other is not null && PathComparer.Equals(Id, other.Id);
    public override bool Equals(object? obj) => obj is FileNode other && Equals(other);
    public override int GetHashCode() => PathComparer.GetHashCode(Id);
    public override string ToString() => DisplayName;

    private static StringComparer PathComparer => StringComparer.Ordinal;
}
